libpromises/dbm_test_api.c

```diff
@@ -0,0 +1,724 @@
+        if (success)
+        {
+            SeqAppend(ret, xstrndup(key, key_size));
+        }
+    }
+    DeleteDBCursor(cur);
+
+    return ret;
+}
+
+/* TODO: LoadDBIntoMap() */
+static Seq *LoadDB(DBHandle *db, size_t limit)
+{
+    DBCursor *cur;
+    bool success = NewDBCursor(db, &cur);
+    if (!success)
+    {
+        return NULL;
+    }
+
+    Seq *ret = SeqNew(16, DBItemDestroy);
+    size_t remaining = limit != 0 ? limit : SIZE_MAX;
+    while (success && (remaining > 0))
+    {
+        int key_size;
+        char *key;
+        int val_size;
+        void *val;
+        success = NextDB(cur, &key, &key_size, &val, &val_size);
+        if (success)
+        {
+            SeqAppend(ret, DBItemNew(key, val_size, val));
+            remaining--;
+        }
+    }
+    DeleteDBCursor(cur);
+
+    return ret;
+}
+
+void DoRandomReads(dbid db_id,
+                   int keys_refresh_s, long min_interval_ms, long max_interval_ms,
+                   bool *terminate)
+{
+    assert(terminate != NULL);
+
+    InitializeRNG((long)gettid());
+
+    DBHandle *db;
+    bool success = OpenDB(&db, db_id);
+    assert(success);
+
+    Seq *keys = GetDBKeys(db);
+    assert(keys != NULL);
+    CloseDB(db);
+
+    size_t n_keys = SeqLength(keys);
+    unsigned long acc_sleeps_ms = 0;
+    while (!*terminate)
+    {
+        long rnd = GetRandomNumber(max_interval_ms - min_interval_ms);
+        unsigned long sleep_for_ms = min_interval_ms + rnd;
+        int ret = usleep(sleep_for_ms * 1000);
+        acc_sleeps_ms += sleep_for_ms;
+        if (ret == -1)
+        {
+            if (errno == EINTR)
+            {
+                continue;
+            }
+            /* else */
+            /* Should never happen. */
+            Log(LOG_LEVEL_ERR, "Failed to usleep() for %ld ms: %s",
+                (min_interval_ms + rnd) * 1000, GetErrorStr());
+            SeqDestroy(keys);
+            return;
+        }
+        rnd = GetRandomNumber(n_keys);
+        unsigned char val[1024];
+        char *key = SeqAt(keys, rnd);
+
+        bool success = OpenDB(&db, db_id);
+        assert(success);
+
+        success = ReadDB(db, key, val, sizeof(val));
+        assert(success || !HasKeyDB(db, key, strlen(key)));
+
+        if (acc_sleeps_ms > (keys_refresh_s * 1000))
+        {
+            SeqDestroy(keys);
+            keys = GetDBKeys(db);
+            assert(keys != NULL);
+            n_keys = SeqLength(keys);
+            acc_sleeps_ms = 0;
+        }
+        CloseDB(db);
+    }
+    SeqDestroy(keys);
+}
+struct ReadParams {
+    dbid db_id;
+    int keys_refresh_s; long min_interval_ms; long max_interval_ms;
+    bool *terminate;
+};
+static void *DoRandomReadsRoutine(void *data)
+{
+    struct ReadParams *params = data;
+    DoRandomReads(params->db_id,
+                  params->keys_refresh_s, params->min_interval_ms, params->max_interval_ms,
+                  params->terminate);
+    return NULL;
```

Comment:
Not imediately clear why this function shouldn't have a return type;

```suggestion
    // Always return NULL, 
    // pthread_create wants a function which returns (void *)
    return NULL;
```